Use go-git for getting the latest commit

raulcabello opened the pull request:
Remove the git shell out command. Use go-git for fetching the latest commit instead.

`InsecureIgnoreHostKey` is used if no known_hosts file is provided, this will accept all hosts.

testcontainers and gogs are used for integration tests. These tests create a git server and test the interaction between gitjob and the git server. The git server is launched in a docker container for each test case (this is handled by testcontainers)

weyfonk commented on integrationtests/git/git_test.go:
```diff
@@ -0,0 +1,374 @@
+)
+
+/*
+This tests uses gogs for testing the integration with a git server. Gogs container uses the data from assets/gitserver, which
```
```suggestion
These tests use gogs for testing integration with a git server. Gogs container uses data from assets/gitserver, which
```

weyfonk commented on integrationtests/git/git_test.go:
```diff
@@ -0,0 +1,374 @@
+		t.Errorf("got error when none was expected: %v", err)
+	}
+	defer func() {
+		if err := container.Terminate(ctx); err != nil {
```
nit: this anonymous container termination function could be returned by `createGogsContainer`, as its body does not change between test cases. I think that this would better isolate container handling logic from test case implementations.

raulcabello replied:
I think it is better to extract it to a separate func, see https://github.com/rancher/gitjob/pull/287/commits/7ce3b9b02d21dfc1295448d079d6041d80249880

weyfonk replied:
That also works for me :+1:

weyfonk commented on integrationtests/assets/gitserver/git/gogs-repositories/test/private-repo.git/description:
```diff
@@ -0,0 +1 @@
+Unnamed repository; edit this file 'description' to name the repository.
```
Just curious: do we _need_ `description` and `exclude` files here?

weyfonk commented on integrationtests/git/git_test.go:
```diff
@@ -0,0 +1,374 @@
+func TestLatestCommit_NoAuth(t *testing.T) {
+	ctx := context.Background()
+	container, url, err := createGogsContainer(ctx, createTempFolder(t))
+	if err != nil {
+		t.Errorf("got error when none was expected: %v", err)
+	}
```
nit: using https://pkg.go.dev/github.com/stretchr/testify/require, this could simply be:
```suggestion
	require.NoError(t, err)
```
(Same remark about `if` blocks throughout this file, checking expectations vs reality, with `stretchr/testify/assert`)

manno replied:
I just wanted to mention `require := require.New(t)` for longer test cases to clean up the arguments to future require calls.

raulcabello replied:
done in https://github.com/rancher/gitjob/pull/287/commits/7ce3b9b02d21dfc1295448d079d6041d80249880

weyfonk commented on integrationtests/git/git_test.go:
```diff
@@ -0,0 +1,374 @@
+					},
+				},
+			},
+			knownHosts:     []byte("doesntexits " + gogsFingerPrint),
```
(very) nit :) :
```suggestion
			knownHosts:     []byte("doesntexist " + gogsFingerPrint),
```

weyfonk commented on integrationtests/git/git_test.go:
```diff
@@ -0,0 +1,374 @@
+const (
+	latestCommitPublicRepo  = "8cd5ab9c851482ce13a544c91ee010f6fdc7cf3f"
+	latestCommitPrivateRepo = "417310891d63d3f3a478bd4c5013e2f532056e8e"
+	gogsFingerPrint         = "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBBpayjxZ7oeeMc6KjGM0VgFEE5GmN1H6RLquUENLcpGcKzrEtym48WmAnX9Xwdkg8eMUBgyYkZtZgR+eapf29fQ="
```
How is this fingerprint generated? Just curious.

manno replied:
`ssh-keygen -lf ./id_ecdsa` I guess? Or ` ssh-keyscan -t ecdsa host`.

weyfonk commented on integrationtests/git/git_test.go:
```diff
@@ -0,0 +1,374 @@
+			secretGetter := &secretGetterMock{secret: secret}
+			latestCommit, err := git.LatestCommit(test.gitjob, secretGetter)
+
+			if !reflect.DeepEqual(err, test.expectedErr) {
```
Do we need equality here, or would [errors.Is](https://pkg.go.dev/errors#Is) do the trick? This could make these test cases more flexible if we decide to enhance error handling with prefixing, wrapping errors with more context across calls.

raulcabello replied:
unfortunately errors.Is fails even if the errors are the same

weyfonk reviewed (commented):
Great work, thanks for this! Using the right tools for the job instead of re-assembling a git server :)
Leaving a few nitpicks and questions.
I'm also wondering why we need this many keys on the server (I may well be missing important details), and whether explicitly committing to each repo as part of the test setup could perhaps be easier to troubleshoot and understand, as opposed to including blobs into this repo.
Could this perhaps make the SQLite DB smaller as well, or is that just wishful thinking? I'm a bit reluctant to push blobs this big to a repo if we can help it.
Happy to discuss all of this :)

raulcabello commented on .github/workflows/integrationtests-ci.yaml:
```diff
@@ -0,0 +1,28 @@
+      - name: integration-test
+        run: go test -v ./integrationtests/...
+        env:
+          TESTCONTAINERS_RYUK_DISABLED: true
```
we need this to run testcontainers in github actions
```
Ryuk must be started as a privileged container.
If your environment already implements automatic cleanup of containers after the execution, but does not allow starting privileged containers, you can turn off the Ryuk container by setting TESTCONTAINERS_RYUK_DISABLED environment variable to true.
```
[More info](https://java.testcontainers.org/features/configuration/)

raulcabello commented:
We could make it smaller by removing the two initial repos (one public and another private). Then we would need to programmatically create them and add commits. 

The size of the whole integrationtests/assets/gitserver folder is small (just 484K), so I don't think it is worth spending time optimizing this. I'm happy to discuss it next week

manno commented:
> The size of the whole integrationtests/assets/gitserver folder is small (just 484K), so I don't think it is worth spending time optimizing this. I'm happy to discuss it next week

I'd prefer the sql `.db` as a dump? Should be possible to `sqlite3 .dump the.db` and restore it with `sqlite3 new.db < the.dump`.

Regarding the git repo in assets, maybe we can store it as a tar.gz?

manno commented on pkg/git/lsremote.go:
```diff
@@ -199,143 +203,52 @@ func (g *git) setCredential(cred *corev1.Secret) error {
+			auth.HostKeyCallback, err = createKnownHosts(cred.Data["known_hosts"])
+		} else {
+			//nolint G106: Use of ssh InsecureIgnoreHostKey should be audited
+			auth.HostKeyCallback = ssh.InsecureIgnoreHostKey()
```
https://fleet.rancher.io/gitrepo-add#adding-private-git-repository has a lot of warnings for private ssh already

manno reviewed (commented):
I'm not sure about all the secret material for the test setup in this PR. The deployment is isolated to the test runner, but I wonder if security scanners will flag our repo?

How difficult would it be to get rid of the secrets and generate them on the fly?

raulcabello commented on pkg/git/lsremote.go:
```diff
@@ -1,28 +1,26 @@
+	httpgit "github.com/go-git/go-git/v5/plumbing/transport/http"
+	gossh "github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	"github.com/go-git/go-git/v5/storage/memory"
 	"github.com/pkg/errors"
```
done in https://github.com/rancher/gitjob/pull/287/commits/7ce3b9b02d21dfc1295448d079d6041d80249880

raulcabello commented:
I removed the harcoded repos in gogs, which has reduced the size of the folder. We now create the repos programatically in each test case. I also removed some unnecessary keys
